server/src/main/java/io/crate/planner/optimizer/iterative/IterativeOptimizer.java

```diff
@@ -0,0 +1,179 @@
+/*
+ * Licensed to Crate.io GmbH ("Crate") under one or more contributor
+ * license agreements.  See the NOTICE file distributed with this work for
+ * additional information regarding copyright ownership.  Crate licenses
+ * this file to you under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.  You may
+ * obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial agreement.
+ */
+
+package io.crate.planner.optimizer.iterative;
+
+import static io.crate.planner.optimizer.Optimizer.removeExcludedRules;
+
+import java.util.List;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.Version;
+
+import io.crate.metadata.CoordinatorTxnCtx;
+import io.crate.metadata.NodeContext;
+import io.crate.planner.operators.LogicalPlan;
+import io.crate.planner.optimizer.Rule;
+import io.crate.planner.optimizer.matcher.Captures;
+import io.crate.planner.optimizer.matcher.Match;
+import io.crate.statistics.TableStats;
+
+/**
+ * The optimizer takes an operator tree of logical plans and creates an optimized plan.
+ * The optimization loop applies rules recursively until a fixpoint is reached.
+ */
+public class IterativeOptimizer {
+
+    private static final Logger LOGGER = LogManager.getLogger(IterativeOptimizer.class);
+
+    private final List<Rule<?>> rules;
+    private final Supplier<Version> minNodeVersionInCluster;
+    private final NodeContext nodeCtx;
+
+    public IterativeOptimizer(NodeContext nodeCtx, Supplier<Version> minNodeVersionInCluster, List<Rule<?>> rules) {
+        this.rules = rules;
+        this.minNodeVersionInCluster = minNodeVersionInCluster;
+        this.nodeCtx = nodeCtx;
+    }
+
+    public LogicalPlan optimize(LogicalPlan plan, TableStats tableStats, CoordinatorTxnCtx txnCtx) {
+        // Create a new memo for the given plan tree
+        var memo = new Memo(plan);
```

Comment:
```suggestion
        var memo = new Memo(plan);
```

Document why vs. what. The comment doesn't explain anything that's not obvious from the code